Spigot-Server-Patches/0419-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,1150 @@
++    }
++
++    @Nullable
++    public ObjectSet<EntityPlayer> getPlayersTracking(long pair) {
++        return this.areaMapTracked.get(pair);
++    }
++
++    @Nullable
++    public ObjectSet<EntityPlayer> getPlayersTracking(int chunkX, int chunkZ) {
++        return this.getPlayersTracking(ChunkCoordIntPair.asLong(chunkX, chunkZ));
++    }
++
++    @Nullable
++    public ObjectSet<EntityPlayer> getPlayersTracking(ChunkCoordIntPair chunkCoordIntPair) {
++        return this.getPlayersTracking(chunkCoordIntPair.x, chunkCoordIntPair.z);
++    }
++
++    public int size() {
++        return this.areaMap.size();
++    }
++    
++    private void attach(long l, EntityPlayer player) {
++        this.areaMap.computeIfAbsent(l, l2 -> new ObjectOpenHashSet<>()).add(player);
++    }
++
++    private void detach(long l, EntityPlayer player) {
++        ObjectSet objectSet = (ObjectSet)this.areaMap.get(l);
++        if(objectSet != null) {
++            objectSet.remove(player);
++            if (objectSet.isEmpty()) this.areaMap.remove(l);
++        }
++    }
++
++    private void attachTrack(long l, EntityPlayer player) {
++        this.areaMapTracked.computeIfAbsent(l, l2 -> new ObjectOpenHashSet<>()).add(player);
++    }
++
++    private void detachTrack(long l, EntityPlayer player) {
++        ObjectSet objectSet = (ObjectSet)this.areaMapTracked.get(l);
++        if(objectSet != null) {
++            objectSet.remove(player);
++            if (objectSet.isEmpty()) this.areaMapTracked.remove(l);
++        }
++    }
++
++}
+diff --git a/src/main/java/net/minecraft/server/ChunkMapDistance.java b/src/main/java/net/minecraft/server/ChunkMapDistance.java
+index 63a688725..191bede32 100644
+--- a/src/main/java/net/minecraft/server/ChunkMapDistance.java
++++ b/src/main/java/net/minecraft/server/ChunkMapDistance.java
+@@ -1,7 +1,6 @@
+ package net.minecraft.server;
+ 
+ import com.google.common.collect.ImmutableList;
+-import com.google.common.collect.Sets;
+ import com.mojang.datafixers.util.Either;
+ import it.unimi.dsi.fastutil.longs.Long2ByteMap;
+ import it.unimi.dsi.fastutil.longs.Long2ByteOpenHashMap;
+@@ -20,7 +19,7 @@ import it.unimi.dsi.fastutil.objects.ObjectIterator;
+ import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
+ import it.unimi.dsi.fastutil.objects.ObjectSet;
+ import it.unimi.dsi.fastutil.objects.ObjectSortedSet;
+-import java.util.Set;
++
+ import java.util.concurrent.CompletableFuture;
+ import java.util.concurrent.Executor;
+ import javax.annotation.Nullable;
+@@ -43,6 +42,11 @@ public abstract class ChunkMapDistance {
+     private final LongSet l = new LongOpenHashSet();
+     private final Executor m;
+     private long currentTick;
++    //Paper start ticket level pp-view-dist
++    protected abstract com.destroystokyo.paper.util.PlayerTicketAreaMap getPlayerTicketAreaMap();
++    protected abstract void updatePlayerTicketAreaMap(EntityPlayer player, long pair);
```

Comment:
This technically breaks ABI (i.e plugins extending this for whatever dumb reason...). There are ways around this, i.e adding another constructor with a callback.

Reply:
just did this in rewrite, added constructor with callback to player chunk map